### PR TITLE
IBM-MQ polling inbound

### PR DESCRIPTION
**Pull request by hasithajayasundara**

### CODE REVIEW

File: src/main/java/org/wso2/carbon/inbound/custom/poll/MQPollingInbound.java [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+* Unless required by applicable law or agreed to in writing,
+* software distributed under the License is distributed on an
+* "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+* KIND, either express or implied.  See the License for the
+* specific language governing permissions and limitations
+* under the License.
+*/
+package org.wso2.carbon.inbound.custom.poll;
+
+import com.ibm.mq.*;
```
Comment (djkevincr) [inline src/main/java/org/wso2/carbon/inbound/custom/poll/MQPollingInbound.java]: Shall we do this imports separately?

**Review by djkevincr — CHANGES_REQUESTED** [review]

File: src/main/java/org/wso2/carbon/inbound/custom/poll/MQPollingInbound.java (hunk 1 above)
Comment (hasithajayasundara) [reply src/main/java/org/wso2/carbon/inbound/custom/poll/MQPollingInbound.java]: Done..!

File: ibmmqpollinginbound.iml [hunk 2]
```diff
@@ -0,0 +1,221 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (djkevincr) [inline ibmmqpollinginbound.iml]: Please remove this intellij idea generated files.

File: org.apache.synapse.custom.poll.class.iml [hunk 3]
```diff
@@ -0,0 +1,218 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<module org.jetbrains.idea.maven.project.MavenProjectsManager.isMavenModule="true" type="JAVA_MODULE" version="4">
```
Comment (djkevincr) [inline org.apache.synapse.custom.poll.class.iml]: Please remove this intellij idea generated files.

File: pom.xml [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+            <id>wso2</id>
+            <name>WSO2 internal Repository</name>
+            <url>http://dist.wso2.org/maven2/</url>
+            <releases>
+                <enabled>true</enabled>
+                <updatePolicy>never</updatePolicy>
+                <checksumPolicy>ignore</checksumPolicy>
+            </releases>
+        </repository>
+
```
Comment (djkevincr) [inline pom.xml]: Please remove this unnecessary new lines.

File: src/main/java/org/wso2/carbon/inbound/custom/poll/MQConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+        }
+
+        if (ibmmqProperties.getProperty(MQConstant.PASSWORD) != null) {
+            this.password = ibmmqProperties.getProperty(MQConstant.PASSWORD);
+        } else {
+            this.password = null;
+        }
+    }
+
+    public int getmaxConnections() {
```
Comment (djkevincr) [inline src/main/java/org/wso2/carbon/inbound/custom/poll/MQConfiguration.java]: Please properly document this public methods for all cases.

File: src/main/java/org/wso2/carbon/inbound/custom/poll/MQConnectionBuilder.java [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * Start connection with IBM MQ queue manager
+ */
+public class MQConnectionBuilder {
```
Comment (djkevincr) [inline src/main/java/org/wso2/carbon/inbound/custom/poll/MQConnectionBuilder.java]: Please make this class a utility class MQConnectionUtils and remove all the class level variable. You need to address all stuff discussed in the IBM-MQ connector PR. [1] Same class is discussed over there as well. 
[1] https://github.com/wso2-incubator/esb-connector-ibmmq/pull/1

File: src/main/java/org/wso2/carbon/inbound/custom/poll/MQConstant.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+* specific language governing permissions and limitations
+* under the License.
+*/
+package org.wso2.carbon.inbound.custom.poll;
+
+/**
+ * IBM MQ constants
+ */
+public class MQConstant {
+
```
Comment (djkevincr) [inline src/main/java/org/wso2/carbon/inbound/custom/poll/MQConstant.java]: Please remove this unnecessary new lines.

File: src/test/java/org/wso2/carbon/inbound/ibmmqintegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+* KIND, either express or implied.  See the License for the
+* specific language governing permissions and limitations
+* under the License.
+*/
+package org.wso2.carbon.inbound;
+
+/**
+ * Created by hasitha on 7/7/17.
+ */
+public class ibmmqintegrationTest {
```
Comment (djkevincr) [inline src/test/java/org/wso2/carbon/inbound/ibmmqintegrationTest.java]: IbmmqintegrationTest

File: src/test/java/org/wso2/carbon/inbound/ibmmqintegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+* software distributed under the License is distributed on an
+* "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+* KIND, either express or implied.  See the License for the
+* specific language governing permissions and limitations
+* under the License.
+*/
+package org.wso2.carbon.inbound;
+
+/**
+ * Created by hasitha on 7/7/17.
```
Comment (djkevincr) [inline src/test/java/org/wso2/carbon/inbound/ibmmqintegrationTest.java]: Please remove this Intellij idea generated comments.

File: src/main/java/org/wso2/carbon/inbound/custom/poll/MQPollingInbound.java [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+                byte[] strData = new byte[strLen];
+                message.readFully(strData);
+
+                //crating new message context
+                msgCtx = this.createMessageContext();
+                msgCtx.setProperty("MessageId", new String(message.messageId));
+                msgCtx.setProperty("CorrelationID", new String(message.correlationId));
+                logger.info("Messsage data-" + new String(strData));
+
+                if (message.getStringProperty("ContentType") != null) {
```
Comment (djkevincr) [inline src/main/java/org/wso2/carbon/inbound/custom/poll/MQPollingInbound.java]: Please check whether the correct message builders are invoked for IBM-MQ content types. Please test this at least for json and xml content types.

File: src/main/java/org/wso2/carbon/inbound/custom/poll/MQPollingInbound.java [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+        logger.info("Removing resources");
+        try {
+            if (manager.isConnected()) {
+                manager.close();
+                manager = null;
+            } else {
+                manager = null;
+            }
+        } catch (MQException e) {
+            logger.error("Error removing services");
```
Comment (djkevincr) [inline src/main/java/org/wso2/carbon/inbound/custom/poll/MQPollingInbound.java]: Exception is dropped from the log error. Please include exception in log.
logger.error("Error removing services", e);

File: src/main/java/org/wso2/carbon/inbound/custom/poll/MQPollingInbound.java [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+                } else {
+                    injectMessage(new String(strData), MQConstant.DEFAULT_CONTENT_TYPE);
+                }
+                readableQueue.close();
+            }
+
+        } catch (Exception e) {
+            logger.info("Queue is empty");
+        }
+
```
Comment (djkevincr) [inline src/main/java/org/wso2/carbon/inbound/custom/poll/MQPollingInbound.java]: Please remove this unnecessary new line.

File: src/main/java/org/wso2/carbon/inbound/custom/poll/MQPollingInbound.java [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+                if (message.getStringProperty("ContentType") != null) {
+                    injectMessage(new String(strData), message.getStringProperty("ContentType"));
+                } else {
+                    injectMessage(new String(strData), MQConstant.DEFAULT_CONTENT_TYPE);
+                }
+                readableQueue.close();
+            }
+
+        } catch (Exception e) {
+            logger.info("Queue is empty");
```
Comment (djkevincr) [inline src/main/java/org/wso2/carbon/inbound/custom/poll/MQPollingInbound.java]: This should be a error log and exception is dropped - logger.error("Message", e);

File: src/main/java/org/wso2/carbon/inbound/custom/poll/MQPollingInbound.java [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+                if (config.getCorrelationID() != null && config.getMessageID() != null) {
+                    gmo.matchOptions = MQConstants.MQMO_MATCH_CORREL_ID + MQConstants.MQMO_MATCH_GROUP_ID;
+                } else if (config.getMessageID() != null) {
+                    gmo.matchOptions = MQConstants.MQMO_MATCH_MSG_ID;
+                } else if (config.getCorrelationID() != null) {
+                    gmo.matchOptions = MQConstants.MQMO_MATCH_CORREL_ID;
+                }
+
+                readableQueue.get(message, gmo);
+                logger.info("A message received");
```
Comment (djkevincr) [inline src/main/java/org/wso2/carbon/inbound/custom/poll/MQPollingInbound.java]: May be this should be a debug log.

**Comment by djkevincr** [comment]
Please note you should remove all the proprietary IBM binary files committed from the PR.
com.ibm.mq.allclient.jar, com.ibm.mq.allclient.jar, com.ibm.mq.allclient.jar

**Comment by djkevincr** [comment]
Current implementation supports polling messages on queue, I think we need to extend this implementation to support topics as well.

File: src/main/java/org/wso2/carbon/inbound/custom/poll/MQConnectionBuilder.java [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+     * Closing IBM-MQ connection
+     */
+    public void closeConnection() {
+        try {
+            if (queueManager.isConnected()) {
+                queueManager.close();
+                queueManager = null;
+            }
+        } catch (MQException e) {
+
```
Comment (hmrajas) [inline src/main/java/org/wso2/carbon/inbound/custom/poll/MQConnectionBuilder.java]: Why this empty catch block?

File: src/main/java/org/wso2/carbon/inbound/custom/poll/MQConnectionBuilder.java [hunk 16]
```diff
@@ -0,0 +1,243 @@
...
+    /**
+     * Get queue manager
+     */
+    public MQQueueManager getManager() {
+        if (queueManager == null || !queueManager.isConnected()) {
+            try {
+                queueManager = new MQQueueManager(config.getqManger());
+            } catch (MQException e) {
+            }
```
Comment (hmrajas) [inline src/main/java/org/wso2/carbon/inbound/custom/poll/MQConnectionBuilder.java]: Empty catch block?

**Pull request closed without merging**